timelock-impl/src/main/java/com/palantir/atlasdb/timelock/watch/LockWatchResourceImpl.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.watch;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.MultimapBuilder;
+import com.google.common.collect.Multimaps;
+import com.palantir.lock.LockDescriptor;
+
+@Path("/lock-watch")
+public class LockWatchResourceImpl implements LockWatchResource {
+    private static final int WATCH_LIMIT = 50;
```

Comment:
is there a way to share between nodes of the same client?

if I've understood correctly, wouldn't this just mean that you have on average `WATCH_LIMIT / num-of-nodes`?

Reply:
continuing based on that assumption, it feels like you would need some sort of additional identifier, similar to reentrancy but for the lock watches, so that different nodes of the same service will get the same watch/id for the same set of things they want to watch

Reply:
we'd definitely need this for client node restarts no? i.e. timelock doesn't go down but the client restarts? or maybe some sort of timeout where the client is also refreshing (you could do this via `getWatchState` I suppose?

Reply:
how do you claim back watches that aren't being observed?

Reply:
1/3 - Not in this version. Yeah, it's possible if you calculate some kind of hash and share IDs along the lines of stream store blocks, or have users explicitly supply an ID.

2 - Yes.

4 - Yes. Having a timeout on watch states seems fine.

5 - Right now there is an explicit un-register endpoint, though it won't help with servers going down. The solution to 4 could call this (there may be more efficient ways of doing it, too)

Reply:
if by version you mean PR that's not consumed by anything, sure, otherwise, I would say it's pretty much required imo as it has api ramifications.